googletest/src/gtest.cc

```diff
@@ -800,41 +810,91 @@ std::string UnitTestImpl::CurrentOsStackTraceExceptTop(int skip_count) {
       );  // NOLINT
 }
 
+// returns the monotonic time in milliseconds
+TimeInMillis GetMonotonicTimeInMillis() {
+#if GTEST_LANG_CXX11
+  const std::chrono::steady_clock::duration now{ std::chrono::steady_clock::now().time_since_epoch() };
+  const std::chrono::milliseconds nowMs{ std::chrono::duration_cast<std::chrono::milliseconds>(now) };
+  return static_cast<TimeInMillis>(nowMs.count());
+#elif GTEST_OS_WINDOWS 
+  static bool freqInit = true; // Note: LARGE_INTEGER is a union so bool needed to avoid punning
+  static LARGE_INTEGER freq;
+  if (freqInit) {
+    freqInit = false;
+    QueryPerformanceFrequency(&freq);
+  }
+  LARGE_INTEGER now;
+  QueryPerformanceCounter(&now);
+  LONGLONG whole = (now.QuadPart / freq.QuadPart);
+  LONGLONG part = (now.QuadPart % freq.QuadPart);
+  return static_cast<TimeInMillis>(whole * 1000 + (part * 1000) / freq.QuadPart);
+#elif GTEST_OS_LINUX
+  timespec now;
+  clock_gettime(CLOCK_MONOTONIC, &now);
+  return static_cast<TimeInMillis>(now.tv_sec) * 1000 + static_cast<TimeInMillis>(now.tv_nsec / 1000000);
+#elif GTEST_OS_MAC
+  static mach_timebase_info_data_t timebaseInfo;
+  if (timebaseInfo.denom == 0) {
+      mach_timebase_info(&timebaseInfo);
+  }
+  // returns time in scaled nanoseconds
+  const unsigned long long now = mach_absolute_time();
+  return static_cast<TimeInMillis>((now / 1000000) * timebaseInfo.numer / timebaseInfo.denom);
+#else
+  // fall back to non monotonic time
+  return GetTimeInMillis();
+#endif
+}
+
 // Returns the current time in milliseconds.
 TimeInMillis GetTimeInMillis() {
-#if GTEST_OS_WINDOWS_MOBILE || defined(__BORLANDC__)
-  // Difference between 1970-01-01 and 1601-01-01 in milliseconds.
+#if GTEST_LANG_CXX11
+  static bool initEpoch = true;
+  static std::chrono::system_clock::time_point epochTime;
```

Comment:
Please check the generated assembly code.
static std::chrono::system_clock::time_point has a non trivial constructor, thus obliging the compiler to generate synchronizing code at the beginning of the function. Ideally epoch time should be outside of the function at the compilation unit level in order to avoid the synchronizing hell.